### PR TITLE
Hot Fix: anonymous donations are now supported across donation types

### LINKED ISSUE #1131: Updating donor profile results in email error
## Issue Overview

The following error seems to prevent any type of update to donor info from the donor profile screen.

Attempting to edit donor information using the `Edit Donor` button results in a server-side error upon submission: `Error: Please enter a valid email address.`

I thought that possibly the email field farther down the page was triggering the error, but entering a valid email in that field still results in the same error. This error occurs when attempting to edit registered and unregistered donors.

## Expected Behavior

User should be able to update donor info from donor profile screen.

## Current Behavior

Clicking `Update Donor` results in an error that references a valid email address. User is left confused because an email field is not part of the form they are updating.

## Possible Solution

Investigate server-side validation to remove the condition causing the error.

## Steps to Reproduce (for bugs)

1. Make a donation with address fields.
2. View donor profile.
3. Click `Edit Donor` button at top of profile.
4. Change address fields.
5. Click `Update Donor`
6. Error appears: `Error: Please enter a valid email address.`

The same error occurs for unregistered users that only have a `User ID` field to edit.

## Screenshots

Error when attempting to update regisered user's address.

image

Error when attempting to update unregisered user's name.

image

## Context

Discovered while investigating address fields for #1112.

## Your Environment

* Give version used: 1.7
* Browser Name and version: Chrome 53.0.2785.143 (64-bit)
* Operating System and version (desktop or mobile): OS X Yosemite 10.10.5

### PR DESCRIPTION
**Pull request by jonwaldstein**
Resolves #1131

## Description

At some point we removed legacy code that updated the anonymous donation meta. This PR adds back support to update the `_give_anonymous_donation` meta that will cover one-time donations using legacy gateways, subscription initial donations, and renewals (legacy subscriptions).  

## Affects

One-time donations (legacy gateways), New subscriptions (initial donation) and renewals (legacy subscriptions)

## Testing Instructions

- Create a subscription with anonymous donation enabled & make sure the subscription _Initial Donation_ reflects this.
- Create a renewal and make sure it's anonymous donation is enabled.
- Check the donor wall for a visual.
- Try donating one-time with "Test Gateway" or any other legacy gateway to make sure anonymous is working.

## Pre-review Checklist

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit and/or end-to-end tests
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

**Comment by jonwaldstein** [comment]
@ravinderk this is ready for review (moved over from give-recurring).  Please don't merge, QA is pending.

File: includes/donors/actions.php [hunk 1]
```diff
@@ -86,3 +90,28 @@ function __give_remove_donor_donation_comment( $donation_id ) {
...
+
+/**
+ * Update anonymous donation for legacy gateways
+ *
+ * @unreleased
+ *
+ * @param $donationId
+ *
+ */
+function updateAnonymousDonationForLegacyGateways($donationId)
```
Comment (ravinderk) [inline includes/donors/actions.php:103]: @jonwaldstein can we add `give` prefix to the function?

File: includes/donors/actions.php [hunk 2]
```diff
@@ -86,3 +90,28 @@ function __give_remove_donor_donation_comment( $donation_id ) {
...
+{
+    $gatewayId = give_get_meta($donationId,'_give_payment_gateway');
```
Comment (ravinderk) [inline includes/donors/actions.php:105]: @jonwaldstein Can we set the third param to `true`. That means the function will return a string value.

File: includes/donors/actions.php [hunk 3]
```diff
@@ -86,3 +90,28 @@ function __give_remove_donor_donation_comment( $donation_id ) {
...
+    $registrar = give(PaymentGatewayRegister::class);
+
+    if (!$registrar->hasPaymentGateway($gatewayId)){
+        $isAnonymousDonation = isset( $_POST['give_anonymous_donation'] ) ? absint( $_POST['give_anonymous_donation'] ) : 0;
+
+        give_update_meta( $donationId, '_give_anonymous_donation', $isAnonymousDonation );
+    }
+}
+
+add_action('give_insert_payment', 'updateAnonymousDonationForLegacyGateways', 10, 2);
```
Comment (ravinderk) [inline includes/donors/actions.php:117]: @jonwaldstein We can remove the third and fourth param.

**Review by ravinderk — CHANGES_REQUESTED** [review]
@jonwaldstein I added a few suggestions. Let me know if you have any questions.

**Comment by jonwaldstein** [comment]
@ravinderk all your comments have been addressed 294b1e95

**Review by ravinderk — CHANGES_REQUESTED** [review]
@jonwaldstein Can you fix failed tests?

**Review by Benunc — APPROVED** [review]
This has been tested by both @rickalday and myself. The only thing this does *not* fix is that existing anonymous donations still show, after updating. 

The fix for that is for users to manually mark those donations as anonymous.

I'm good to merge it.

**Comment by JasonTheAdams** [comment]
@ravinderk Please approve this if it looks good to you and we'll get it out tomorrow.

**Review by ravinderk — APPROVED** [review]

**Pull request merged**